### DISCUSSION TITLE
Exposing Prometheus metrics in rootless mode

### DISCUSSION THREAD

**Discussion by leonboot** [moby/moby#47825] [Q&A]
I'm trying to expose the metrics endpoint (`metrics-addr` configuration option) so it can be scraped by a metrics collector, but I'm running into a snag which is related to Docker running in rootless mode.

I have the following configuration in ~/.config/docker/daemon.json:

```json
{
    "metrics-addr": "localhost:9323"
}
```

Since Docker runs in rootless mode, networking is done in userland (using `slirp4netns`, if I'm correct). It seems the metrics endpoint is also being exposed this way, since running `curl http://localhost:9323/metrics` results in a connection refused error:

```
$ curl http://localhost:9323/metrics
curl: (7) Failed to connect to localhost port 9323 after 0 ms: Couldn't connect to server
```

When I run the command in a Docker container with its network mode set to `host`, I get results:
```
$ docker run --rm -ti --network host alpine ash
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
Digest: sha256:c5b1261d6d3e43071626931fc004f70149baeba2c8ec672bd4f27761f8e1ad6b
Status: Downloaded newer image for alpine:latest
/ # apk --no-cache add curl
fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/community/x86_64/APKINDEX.tar.gz
(1/8) Installing ca-certificates (20240226-r0)
(2/8) Installing brotli-libs (1.1.0-r1)
(3/8) Installing c-ares (1.27.0-r0)
(4/8) Installing libunistring (1.1-r2)
(5/8) Installing libidn2 (2.3.4-r4)
(6/8) Installing nghttp2-libs (1.58.0-r0)
(7/8) Installing libcurl (8.5.0-r0)
(8/8) Installing curl (8.5.0-r0)
Executing busybox-1.36.1-r15.trigger
Executing ca-certificates-20240226-r0.trigger
OK: 12 MiB in 23 packages
/ # curl http://localhost:9323/metrics
# HELP builder_builds_failed_total Number of failed image builds
# TYPE builder_builds_failed_total counter
builder_builds_failed_total{reason="build_canceled"} 0
builder_builds_failed_total{reason="build_target_not_reachable_error"} 0
builder_builds_failed_total{reason="command_not_supported_error"} 0
[...]
```

So the metrics server works, but it is not reachable by the metrics collector which runs on the host OS.

Does anybody know of a way to have the metrics service exposed so the metrics collector can connect to it?

**Comment by AkihiroSuda** [comment]
Will be fixed in:
- https://github.com/moby/moby/pull/47103
